hasktorch/src/Torch/Typed/Device.hs

```diff
@@ -83,3 +112,123 @@ instance (GHasToDevice device' device f g) => GHasToDevice device' device (M1 i
+
+class HasToDevices
+  (devices' :: [(D.DeviceType, Nat)])
+  (devices  :: [(D.DeviceType, Nat)])
+  (fs :: [Type])
+  (gs :: [Type]) | devices' devices fs -> gs, devices' devices gs -> fs where
+  toDevices :: HList fs -> HList gs
+
+instance HasToDevices '[] '[] '[] '[] where
+  toDevices HNil = HNil
+
+instance
+  ( HasToDevices devices' devices fs gs
+  , HasToDevice device' device f g
+  ) => HasToDevices (device' ': devices') (device ': devices) (f ': fs) (g ': gs) where
+  toDevices (f :. fs) = Torch.Typed.Device.toDevice @device' @device f :. toDevices @devices' @devices @fs @gs fs
+
+-- >>> :kind! GetDevice (Torch.Typed.NN.Linear 1 1 'D.Float '( 'D.CPU, 0))
+-- GetDevice (Torch.Typed.NN.Linear 1 1 'D.Float '( 'D.CPU, 0)) :: Maybe (D.DeviceType, Nat)
+-- = 'Just '( 'D.CPU, 0)
+--
+-- >>> :kind! GetDevice (Torch.Typed.Tensor.Tensor '( 'D.CUDA, 0) 'D.Float '[1])
+-- GetDevice (Torch.Typed.Tensor.Tensor '( 'D.CUDA, 0) 'D.Float '[1]) :: Maybe (D.DeviceType, Nat)
+-- = 'Just '( 'D.CUDA, 0)
+type family GetDevice (f :: k) :: Maybe (D.DeviceType, Nat) where
+  GetDevice (t (device :: (D.DeviceType, Nat))) = Just device
+  GetDevice (t a)                               = GetDevice t
+  GetDevice t                                   = Nothing
+
+-- >>> :kind! GetDevices '[Torch.Typed.Tensor.Tensor '( 'D.CUDA, 0) 'D.Float '[1], Torch.Typed.Tensor.Tensor '( 'D.CUDA, 1) 'D.Float '[1]]
+-- GetDevices '[Torch.Typed.Tensor.Tensor '( 'D.CUDA, 0) 'D.Float '[1], Torch.Typed.Tensor.Tensor '( 'D.CUDA, 1) 'D.Float '[1]] :: [(D.DeviceType, Nat)]
+-- = '[ '( 'D.CUDA, 0), '( 'D.CUDA, 1)]
+type family GetDevices (fs :: [k]) :: [(D.DeviceType, Nat)] where
+  GetDevices '[]       = '[]
+  GetDevices (f ': fs) = MaybePrepend (GetDevice f) (GetDevices fs)
+
+-- class HasChunk chunks f gs | chunks f -> gs where
+--   chunk :: f -> HList gs
+
+-- class GHasChunk
+--   (chunks :: Nat)
+--   (f :: Type -> Type)
+--   (gs :: [Type]) | chunks f -> gs where
+--   gChunk :: forall a . f a -> HList gs
+
+-- class GZipChunks (gs :: [k]) (gs' :: [k]) (gs'' :: [k]) | gs gs' -> gs'' where
+--   gZipChunks   :: HList gs -> HList gs' -> HList gs''
+--   -- gUnzipChunks :: HList gs'' -> HList gs :*: HList gs'
+
+-- instance GZipChunks '[] '[] '[] where
+--   gZipChunks _ _ = HNil
+--   -- gUnzipChunks _ = HNil :*: HNil
+
+-- instance
+--   ( (g :*: g') ~ g''
+--   , GZipChunks gs gs' gs''
+--   ) => GZipChunks (g ': gs) (g' ': gs') (g'' ': gs'') where
+--   gZipChunks (g :. gs) (g' :. gs') = (g :*: g') :. gZipChunks gs gs'
+  -- gZipChunks x y = _undefined
+  -- gUnzipChunks (~(g :*: g') :. gs'') =
+  --   let ~(gs :*: gs') = gUnzipChunks gs''
+  --   in  (g :. gs') :*: (y :. ys)
+
+-- instance
+--   (
+
+--   ) => 
+
+-- class HasCat fs g | fs -> g where
+--   cat :: HList fs -> g
+
+class HasScatter devices' device f gs | devices' device f -> gs where
+  scatter :: f -> HList gs
+
+instance
+  ( chunks ~ ListLength devices'
```

Comment:
This property requires the number of devices to be the same length as input-chunks.

Reply:
yes, that's true

Reply:
When the tensor's size is 100,   the size of tensor is 50 as follows.

-- >>> :type scatter @'[ '( 'D.CUDA, 0), '( 'D.CUDA, 1)] (ones @'[100,1] @'D.Float @'( 'D.CPU, 0))
  -- scatter @'[ '( 'D.CUDA, 0), '( 'D.CUDA, 1)] (ones @'[100,1] @'D.Float @'( 'D.CPU, 0))
  --  :: HList
  --       '[Tensor '( 'D.CUDA, 0) 'D.Float '[50, 1],
  --         Tensor '( 'D.CUDA, 1) 'D.Float '[50, 1]]

Reply:
If the tensor's size is 101, Are the sizes of tensor 50 and 51?

Reply:
yes, as defined by the `Chunk` type class, see https://github.com/hasktorch/hasktorch/blob/9b576b67bff54badf3fee8e470f6c652f832923c/hasktorch/src/Torch/Typed/Functional.hs#L1712 for the implementation.

there are cases in which `chunk` returns fewer chunks than requested, for instance, when you try to split a tensor of `'[2]` into 3 pieces. in this case, though, this instance wouldn't be available, because `chunks ~ ListLength devices'` would not hold.